### ISSUE TITLE
azurerm_key_vault not setting purge protection and error when attempting to purge keys.

### ISSUE THREAD

**Issue by jwshive** [hashicorp/terraform-provider-azurerm#9984]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version
I am using azurerm provider 2.41.0 and terraform binary 0.14.3

### Affected Resource(s)

* `azurerm_key_vault`

### Terraform Configuration Files

```hcl
resource "azurerm_key_vault" "key_vault" {
  location = var.resource-location
  name = "${var.agency-name}-${var.department-name}-${var.application-name}-${var.environment}-kv"
  resource_group_name = "zus1-${var.agency-name}-${var.application-name}-${var.environment}-v1-rg"
  sku_name = var.key-vault-sku-name
  tenant_id = data.azurerm_client_config.current_client_config.tenant_id
  soft_delete_enabled = false
  purge_protection_enabled = false
  tags = merge(local.common_tags, map("type", "key-vault"))
}
```

### Debug Output

Error: purging Secret "sql-1-password" (Key Vault "https://my-key-vault.vault.azure.net/"): keyvault.BaseClient#PurgeDeletedSecret: Failure responding to request: StatusCode=400 -- 
Original Error: autorest/azure: Service returned an error. Status=400 Code="NotSupported" Message="Operation \"purge\" is not enabled for this vault."

### Expected Behaviour

I expect the key vault to be created with enablePurgeProtection: false and a terraform destroy to not error out when removing a key vault.

### Actual Behaviour
The key vault is created, but purge protection is null and not false.
...
    "enablePurgeProtection": null,
    "enableRbacAuthorization": false,
    "enableSoftDelete": false,
    "enabledForDeployment": false,
    "enabledForDiskEncryption": false,
    "enabledForTemplateDeployment": false,
....

A terraform destroy results in the error in debug output.
### Steps to Reproduce

1. `terraform destroy`

### Important Factoids

This was working last week as of Thursday or Friday.

### References
This may be related to 

* #5659

**Comment by esciara** [comment]
Same here. When creating a key vault using the defaults, creating a secret, and then destroying the secret, the above mentioned error occurs. 

with `azurerm` version `2.41.0`.

**Comment by ruandersMSFT** [comment]
I'll have to run a test to know if it is a permission problem or is in relation to the Key Vault itself, but for consideration be aware a breaking change was recently enforced in which [Soft Delete is now required to be enabled](https://docs.microsoft.com/en-us/azure/key-vault/general/soft-delete-change).

**Comment by jwshive** [comment]
I gave my service principal the purge permission as documented [here](https://docs.microsoft.com/en-us/azure/key-vault/general/soft-delete-overview) and I still get the error posted in the opening notes.

My key vault sp permissions creation code is now

```hcl
resource "azurerm_key_vault_access_policy" "sp-key_vault_access_policy" {
  key_vault_id = azurerm_key_vault.key_vault.id
  object_id = data.azurerm_client_config.current_client_config.object_id
  tenant_id = data.azurerm_client_config.current_client_config.tenant_id
  key_permissions = ["get", "list", "create", "encrypt", "delete", "purge"]
  secret_permissions = ["get", "list", "set", "delete", "purge"]
  depends_on = [azurerm_key_vault.key_vault]
}
```

**Comment by jackofallops** [comment]
Hi all - just looking at a possibly related issue and thought I should chip in with a couple things:

`purge_protection_enabled` - should only be set to `true`, the API doesn't accept `false`, so we don't send it (so will always be `null` / `nil` if not `true`). If Purge Protection is not desired, please omit this property completely (note: Once set to `true`, it cannot be removed or set to `false`)

`soft_delete_enabled` can now only be `true` as of December 31st 2020, and defaults to this value regardless of what is set in Terraform - we're looking at either accepting 9988, or deprecating the property altogether (and ignoring any value set) - depending on the behaviours of the Azure environments other than Public (it's high on our list to look at).

The behaviour for controlling whether a Key Vault item (Key / Secret etc) is purged when deleted is controlled by the `features` block in the provider config. e.g.
```
provider "azurerm" {
  features {
    key_vault {
      purge_soft_delete_on_destroy = false
    }
  }
}
```
This setting defaults to `true`, so if you do not wish for deleted items to be purged, this can be set to `false`, which will allow the recovery of these items. 

The 409 Conflict is of interest, @jwshive is that consistent? We explicitly check the item is deleted before attempting to purge it, so this may indicate a bug or inconsistency between the read and purge operations.

**Comment by peteneville** [comment]
Same for me: 
Error: purging Certificate "91-cert" (Key Vault "https://91-uks.vault.azure.net/"): keyvault.BaseClient#PurgeDeletedCertificate: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="NotSupported" Message="Operation \"purge\" is not enabled for this vault."

Using azurem 2.41.0 and not setting purge_protection_enabled (so should have default of false). 

Cert was removed though in above run (not unexpected due to code change) and subsequent attempt re-created it.

**Comment by U-bit-solo** [comment]
My KeyVault has SoftDelete and Purge Protection enabled. But I am getting this error, using a Service Principal after updating to version terraform-provider-azurerm: 2.43.0. The config. file was working fine before
Error: purging Key "keystxxxxx" (Key Vault "https://akvteu2nubep40.vault.azure.net/"): keyvault.BaseClient#PurgeDeletedKey: Failure responding to request: StatusCode=403 Code="Forbidden" Message="The user, group or application 'appid=5xxxxxxx;oid=xxxxxxx-aa5cxxxx;numgroups=1;iss=https://sts.windows.net/xxxxxb7e-28xxxxb/' does not have keys purge permission on key vault 'akvteu2xxxxx;location=eastus2'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287" InnerError={"code":"ForbiddenByPolicy"}

**Comment by tommydejong** [comment]
I am getting the same error on azurerm 2.43.0. I'm using a module to deploy virtual machines in Azure using GitLab CI and a service principal. I used to store the storage account access key in the module, but I've removed that feature. When I try to run the module again it attempts to delete the secret in the key vault (not the key vault itself), I get the following:

Error: purging Secret "sa-access-key" (Key Vault "https://...vault.azure.net/"): keyvault.BaseClient#PurgeDeletedSecret: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="Forbidden" Message="The user, group or application 'appid=xxx;oid=xxx;iss=https://sts.windows.net/xxx/' does not have secrets purge permission on key vault '...;location=westeurope'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287" InnerError={"code":"ForbiddenByPolicy"}

This is even though I am certain the service principal has the right access to the key vault:
![image](https://user-images.githubusercontent.com/70956477/106457492-58550580-648f-11eb-9ea6-c9a44ef0f6b5.png)

**Comment by omerfsen** [comment]
+1 I get it too with terraform 0.14.6 and azure provider 2.46.1

**Comment by jason-johnson** [comment]
> The 409 Conflict is of interest, @jwshive is that consistent? We explicitly check the item is deleted before attempting to purge it, so this may indicate a bug or inconsistency between the read and purge operations.

@jackofallops 

For us this is very consistent since this morning at least.  Create environment, then destroy and we get this error.  If we simply run destroy again we get a permission error (presumably the policy access was deleted successfully) and running a 3rd and all consecutive times reports the key vault cannot be found.  Rerun create again and then destroy and the 409 error comes back.

So to summarise, the first destroy we run always gets a 409 on the key vault secrets now.  We are running azurerm 2.47.0 and terraform 0.14.7.  But I bumped the versions hoping it would fix it. This morning we were on azurerm 2.44.0 and terraform 0.14.4 with the exact same problem.

**Comment by tommydejong** [comment]
My previous problem seems to be resolved after upgrading to terraform 0.14.7 and azurerm 2.48.0, but I'm experiencing the 409 error now as well.

**Comment by schlbra** [comment]
Our pipelines are dead in the water due to this bug and our heavy reliance on azure key vault. :(

**Comment by davenicoll** [comment]
I just ran into this issue too (azurerm 2.50 terraform 0.13.9). My keyvaults have soft delete enabled. Running a destroy successfully deletes the secrets, but throws a 403 Forbidden when trying to purge them and ultimately fails. If I re-run terraform, instead of creating the secret with the values of the applied plan, the deleted secrets are restored in the keyvault with their previous values. Bonkers.

**Comment by mattduguid** [comment]
Using terraform 0.14.6 and azurerm v2.5.0.

On recent attempts to destroy we received following error when it tries to delete a keyvault, the weird bit is that it does manage to delete the keys just not the keyvault itself,

Error: Error purging of Secret "<SECRET_NAME>" (Key Vault "https://<KEYVAULT_NAME>.vault.azure.net/") : keyvault.BaseClient#PurgeDeletedSecret: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="NotSupported" Message="Operation \"purge\" is not enabled for this vault."

Similar to @tommydejong we added following and our http/400 issue is now resolved,

provider "azurerm" {
  ...
  features {
    key_vault {
      **purge_soft_delete_on_destroy = false** # https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs#purge_soft_delete_on_destroy
      **recover_soft_deleted_key_vaults = false** # https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs#recover_soft_deleted_key_vaults
    }
  }
  ...
}

**Comment by mattduguid** [comment]
@daguito81 have you tried with both set to false in the provider and false in the resource from a brand new environment,

provider "azurerm" {
  ...
  features {
    key_vault {
        recover_soft_deleted_key_vaults = false
        purge_soft_delete_on_destroy = false
    }
  }
 ...
}

resource "azurerm_key_vault" "azure" {
  ...
  purge_protection_enabled = false _# once enabled cannot be unset, if enabled and you delete keyvault then it remains soft deleted for value of soft_delete_retention_days before name can be reused which causes problems for terraform if you delete and create an environment quickly as name cannot be reused_
  soft_delete_retention_days = 7 # _set lowest possible just in case parameter purge_protection_enabled is set to true_
 ...
}

also this article has some information on soft delete/purge protection for azure keyvaults -> https://docs.microsoft.com/en-us/azure/key-vault/general/soft-delete-change

**Comment by daguito81** [comment]
@mattduguid Yeah, i tested it with both and the behaviour was still funy. I did some tests just now. And terraform is deploying the KeyVault the right way. The problem was basically that attaching a KeyVault to an Azure Machine Learning workspace, automatically activates Purge Protection on that Keyvault. So Terraform deploys without Purge,  then AzureML turns on the purge and then terraform tries to turn purge protection from true to nil, and it can't do that because purge protection can't be deactivated. 

Then when destroying the environment, if you have the option "purge deleted keyvaults" set to true. Then AzureML (and maybe other attached services have this behaviour as well) turns on purge protection, terraform tries to purge on delete, and everything fails. 

I'm trying to figure out why AzureML is activating purge protection on the attached key vault.

**Comment by bitsofinfo** [comment]
same problem: using `2.54`

The KV was created with `purge_protection_enabled = false`, then on another run. It says purge protection is enabled

```
Error: updating Vault: (Name "my-kv" / Resource Group "my-rg"): once Purge Protection has been Enabled it's not possible to disable it
```

```
terraform {
  required_version = "= 0.14.9"

  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = ">= 2.54"
    }
  }
}


data "azurerm_client_config" "current" {}


resource "azurerm_key_vault" "kv" {
  name                = "${var.name_prefix}-kv"
  location            = var.location
  resource_group_name = var.resource_group_name
  tenant_id           = data.azurerm_client_config.current.tenant_id

  sku_name                  = "standard"
  enable_rbac_authorization = true

  enabled_for_disk_encryption     = false
  enabled_for_template_deployment = false
  enabled_for_deployment          = false
  
  purge_protection_enabled = false
  soft_delete_retention_days = 30


  network_acls {
    bypass                     = "AzureServices"
    default_action             = "Deny"
    ip_rules                   = var.network_acl_ip_rules
    virtual_network_subnet_ids = var.network_acl_virtual_network_subnet_ids
  }

  tags = var.tags

}
```

**Comment by PhilippeVandeVyverInetumRealdolmen** [comment]
Has someone found a solution for this?

**Comment by wahyuen** [comment]
For those following this issue, we ran into the same problem (not in Terraform, but using Bicep). After a few back and forth tickets, it appears they have now rolled out a change so that linking a KV to an ML workspace via ARM does NOT auto activate purge protection.

I have tested this in `australiaeast` region, YMMV depending what region you are doing this to.

From the support ticket we raised
> We are on our way to make purge protection not required when creating ML workspace. 

Hopefully this helps some of you with your own use cases!

**Comment by tombuildsstuff** [comment]
👋 

Since this has been fixed in the upstream API I'm going to close this issue out - #10273 is tracking support for more granular purge protection for resources, as such if your interested in that would you mind subscribing to that issue for updates?

Thanks!
